mongo/integration/client_test.go

```diff
@@ -768,6 +768,88 @@ func TestClient(t *testing.T) {
 				"expected 'OP_MSG' OpCode in wire message, got %q", pair.Sent.OpCode.String())
 		}
 	})
+
+	opts := mtest.NewOptions().
+		// Blocking failpoints don't work on pre-4.2 and sharded clusters.
+		Topologies(mtest.Single, mtest.ReplicaSet).
+		MinServerVersion("4.2").
+		// Expliticly enable retryable reads and retryable writes.
+		ClientOptions(options.Client().SetRetryReads(true).SetRetryWrites(true))
+	mt.RunOpts("operations don't retry after a context timeout", opts, func(mt *mtest.T) {
+		testCases := []struct {
+			desc      string
+			failPoint mtest.FailPoint
+			operation func(context.Context, *mongo.Collection) error
+		}{
+			{
+				desc: "read op",
+				failPoint: mtest.FailPoint{
+					ConfigureFailPoint: "failCommand",
+					Mode:               "alwaysOn",
+					Data: mtest.FailPointData{
+						FailCommands:    []string{"find"},
+						BlockConnection: true,
+						BlockTimeMS:     500,
+					},
+				},
+				operation: func(ctx context.Context, coll *mongo.Collection) error {
+					return coll.FindOne(ctx, bson.D{}).Err()
+				},
+			},
+			{
+				desc: "write op",
+				failPoint: mtest.FailPoint{
+					ConfigureFailPoint: "failCommand",
+					Mode:               "alwaysOn",
+					Data: mtest.FailPointData{
+						FailCommands:    []string{"insert"},
+						BlockConnection: true,
+						BlockTimeMS:     500,
+					},
+				},
+				operation: func(ctx context.Context, coll *mongo.Collection) error {
+					_, err := coll.InsertOne(ctx, bson.D{})
+					return err
+				},
+			},
+		}
+
+		for _, tc := range testCases {
+			mt.Run(tc.desc, func(t *mtest.T) {
+				_, err := mt.Coll.InsertOne(context.Background(), bson.D{})
+				require.NoError(mt, err)
+
+				mt.SetFailPoint(tc.failPoint)
+
+				mt.ClearEvents()
+
+				for i := 0; i < 50; i++ {
+					// Run 50 operations, each with a timeout of 50ms. Expect
+					// them to all return a timeout error because the failpoint
+					// blocks find operations for 500ms. Run 50 to increase the
+					// probability that an operation will time out in a way that
+					// can cause a retry.
+					ctx, cancel := context.WithTimeout(context.Background(), 50*time.Millisecond)
+					defer cancel()
```

Comment:
```suggestion
					mt.Cleanup(cancel)
```

nit. suggest using Cleanup over defer in tests

Reply:
Interesting idea. I think in this case, deferring a call in a loop is actually not a good practice, so I'll update to explicitly call `cancel` after the operation returns.